test/common/stats/thread_local_store_test.cc

```diff
@@ -90,6 +90,115 @@ class StatsThreadLocalStoreTest : public testing::Test, public RawStatDataAlloca
   std::unique_ptr<ThreadLocalStoreImpl> store_;
 };
 
+class HistogramTest : public testing::Test, public RawStatDataAllocator {
+public:
+  void SetUp() override {
+    InSequence s;
+    ON_CALL(*this, alloc(_)).WillByDefault(Invoke([this](const std::string& name) -> RawStatData* {
+      return alloc_.alloc(name);
+    }));
+
+    ON_CALL(*this, free(_)).WillByDefault(Invoke([this](RawStatData& data) -> void {
+      return alloc_.free(data);
+    }));
+
+    EXPECT_CALL(*this, alloc("stats.overflow"));
+    store_.reset(new ThreadLocalStoreImpl(*this));
+    store_->addSink(sink_);
+    store_->initializeThreading(main_thread_dispatcher_, tls_);
+  }
+
+  void TearDown() override {
+    store_->shutdownThreading();
+    tls_.shutdownThread();
+    // Includes overflow stat.
+    EXPECT_CALL(*this, free(_));
+  }
+
+  std::vector<uint64_t> h1_cumulative_values, h2_cumulative_values, h1_interval_values,
+      h2_interval_values;
+
+  /**
+   * Validates that Histogram merge happens as desired and returns the processed histogram count
+   * that can be asserted later.
+   */
+  uint64_t validateMerge() {
+    std::shared_ptr<std::atomic<bool>> merge_called = std::make_shared<std::atomic<bool>>(false);
+    store_->mergeHistograms([merge_called]() -> void { *merge_called = true; });
+
+    EXPECT_TRUE(*merge_called);
+
+    std::list<ParentHistogramSharedPtr> histogram_list = store_->histograms();
+
+    histogram_t* hist1_cumulative = hist_alloc();
+    for (uint64_t value : h1_cumulative_values) {
+      hist_insert_intscale(hist1_cumulative, value, 0, 1);
+    }
+
+    histogram_t* hist2_cumulative = hist_alloc();
+    for (uint64_t value : h2_cumulative_values) {
+      hist_insert_intscale(hist2_cumulative, value, 0, 1);
+    }
+
+    histogram_t* hist1_interval = hist_alloc();
+    for (uint64_t value : h1_interval_values) {
+      hist_insert_intscale(hist1_interval, value, 0, 1);
+    }
+
+    histogram_t* hist2_interval = hist_alloc();
+    for (uint64_t value : h2_interval_values) {
+      hist_insert_intscale(hist2_interval, value, 0, 1);
+    }
+
+    HistogramStatisticsImpl h1_cumulative_statistics(hist1_cumulative);
+    HistogramStatisticsImpl h2_cumulative_statistics(hist2_cumulative);
+    HistogramStatisticsImpl h1_interval_statistics(hist1_interval);
+    HistogramStatisticsImpl h2_interval_statistics(hist2_interval);
+
+    for (const Stats::ParentHistogramSharedPtr& histogram : histogram_list) {
+      if (histogram->name() == "h1") {
```

Comment:
OK this pattern of looping over the histograms is a bit awkward and I think it might be worth having a test-method:

```
typedef std::map<std::string, Stats::ParentHistogramSharedPtr> NameHistogramMap
NameHistogramMap makeHistogramMap(list) { ... };
```

with usage in tests
```
const Stats::ParentHistogramSharedPtr& h1 = histogram_map["h1"];
EXPECT_EQ(h1->cumulativeStatistics().summary(), h1_cumulative_statistics.summary());
...
```
rather than looping and if/elseing the cases.  If it were just one of these then it might not be worth it, but there are several, here & below.

Reply:
done